src/main/java/com/salesforce/dynamodbv2/mt/util/CachingAmazonDynamoDbStreams.java

```diff
@@ -529,6 +568,96 @@ private CachingAmazonDynamoDbStreams(AmazonDynamoDBStreams amazonDynamoDbStreams
         GuavaCacheMetrics.monitor(meterRegistry, iteratorCache, cn + ".GetShardIterator");
     }
 
+    /**
+     * Gets the {@code DescribeStreamResult} from cache or from DescribeStream API.
+     *
+     * @return Stream details for the given request (streamArn).
+     */
+    private DescribeStreamResult lookupDescribeStreamResult(DescribeStreamRequest request)
+        throws AmazonDynamoDBException {
+
+        String key = request.getStreamArn();
+        try {
+            DescribeStreamResult result = describeStreamCache.get(key,
+                () -> this.loadStreamDescriptionForAllShards(request));
+            return result;
+        } catch (UncheckedExecutionException | ExecutionException | InvalidCacheLoadException e) {
+            // Catch exceptions thrown on cache lookup or cache loader and try load method once more.
+            // (get call will throw UncheckedExecutionException before aws exception (i.e. AmazonDynamoDBException)).
+            LOG.warn("Failed getting DescribeStreamResult for all shards from cache lookup or load method.  "
+                + "Retrying describeStream call.");
+            e.printStackTrace();
+            return this.loadStreamDescriptionForAllShards(request);
+        }
+    }
+
+    /**
+     * Gets the {@code DescribeStreamResult} from the DescribeStream API.
+     *
+     * @return Stream details for the given request with all the shards currently available.
+     */
+    protected DescribeStreamResult loadStreamDescriptionForAllShards(DescribeStreamRequest request)
+        throws AmazonDynamoDBException {
+
+        List<Shard> allShards = new ArrayList<>();
+        String streamArn = request.getStreamArn();
+        String lastShardId = request.getExclusiveStartShardId();
+
+        if (LOG.isDebugEnabled()) {
+            LOG.debug("cache miss in describe stream cache for key: {}", streamArn);
+        }
+        // TODO metrics: count cache miss
+
+        DescribeStreamResult result = null;
+        DescribeStreamResult previousResult;
+        do {
+            request = new DescribeStreamRequest().withStreamArn(streamArn).withExclusiveStartShardId(lastShardId);
+            previousResult = result;
+            result = super.describeStream(request);
+
+            if (result != null && result.getStreamDescription().getShards() != null) {
```

Comment:
Is result ever null?

Reply:
In Kinesis Proxy when calling `describeStream`, it throws an exception if it's null after retrying and getting a null response: https://github.com/awslabs/amazon-kinesis-client/blob/09ec52bd16e5293cb06e53c05089b76adb30a76f/src/main/java/com/amazonaws/services/kinesis/clientlibrary/proxies/KinesisProxy.java#L276

I don't think the result will be null though by looking through the aws sdk.  If there's an error it should throw the known exceptions (which mt-dynamo clients should expect) - http://docs.aws.amazon.com/goto/WebAPI/kinesis-2013-12-02/DescribeStream.  Looking through the aws sdk, it looks like it returns a non null response since it calls response.getAwsResponse(); and adds an errorResponseHandler to make the response non-null if there is an error.